File: electron-app/src/services/llmService.js
/* ═══════════════════════════════════════════════════════════════════════════
   LLMService — Polish, translate, or summarize transcripts via OpenAI-compatible API
   ═══════════════════════════════════════════════════════════════════════════ */

class LLMService {
  getSettings() {
    return {
      endpoint: localStorage.getItem('llm_endpoint') || 'https://api.groq.com/openai/v1',
      apiKey: localStorage.getItem('llm_api_key') || localStorage.getItem('groq_api_key') || '',
      model: localStorage.getItem('llm_model') || 'llama-3.3-70b-versatile',
    };
  }

  async process(text, mode, targetLang) {
    const { endpoint, apiKey, model } = this.getSettings();
    if (!apiKey) throw new Error('LLM API key not set. Go to Settings → configure LLM.');

    const systemPrompts = {
      'polish': 'You are a transcription editor. Fix grammar, punctuation, and awkward phrasing. Preserve the original meaning, speaker intent, and tone. Remove filler words (um, uh, like) where appropriate. Output only the polished text, no commentary.',
      'translate': `You are a translator. Translate the following text to ${targetLang}. Preserve the meaning, tone, and style. For ${targetLang === 'Greek' ? 'Greek, use the Greek alphabet (ελληνικά).' : ''} Output only the translation, no commentary.`,
      'summarize': 'Provide a concise summary of the following transcript. Capture the key points, decisions, and action items. Output in clear bullet points.',
      'fix-punctuation': 'Add proper punctuation (periods, commas, question marks, capitalization) to this transcript. Do NOT change any words or their order. Output only the punctuated text.',
    };

    const prompt = systemPrompts[mode];
    if (!prompt) throw new Error(`Unknown mode: ${mode}`);

    const url = `${endpoint.replace(/\/$/, '')}/chat/completions`;
    const resp = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model,
        messages: [
          { role: 'system', content: prompt },
          { role: 'user', content: text },
        ],
        temperature: 0.3,
        max_tokens: 4096,
      }),
    });

    if (!resp.ok) {
      let errMsg = `HTTP ${resp.status}`;
      try {
        const err = await resp.json();
        errMsg = err.error?.message || errMsg;
      } catch {}
      throw new Error(errMsg);
    }

    const data = await resp.json();
    return data.choices?.[0]?.message?.content?.trim() || '';
  }
}

window.llmService = new LLMService();

File: Sources/TranscriptionApp/LLMService.swift
import Foundation
import AppKit

// MARK: - LLM Service (OpenAI-Compatible API)
final class LLMService {
    static let shared = LLMService()

    private let session: URLSession
    private let decoder = JSONDecoder()

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 60
        config.timeoutIntervalForResource = 120
        self.session = URLSession(configuration: config)
    }

    private var baseURL: String {
        UserDefaults.standard.string(forKey: SettingsKeys.llmBaseURL) ?? "https://api.groq.com/openai/v1"
    }

    private var model: String {
        UserDefaults.standard.string(forKey: SettingsKeys.llmModel) ?? "llama-3.3-70b-versatile"
    }

    private var apiKey: String {
        UserDefaults.standard.string(forKey: SettingsKeys.llmApiKey) ?? ""
    }

    // MARK: - Public API

    func polish(text: String) async throws -> String {
        let systemPrompt = """
        You are a transcription polishing assistant. Fix small errors in transcribed text —
        spelling mistakes, missing punctuation, incorrect words that sound similar (homophones),
        and formatting issues. Keep the SAME LANGUAGE as the input. Do NOT rephrase, summarize,
        or change the meaning. Only fix obvious transcription errors.
        Return ONLY the corrected text, nothing else.
        """
        return try await chatCompletion(systemPrompt: systemPrompt, userMessage: text)
    }

    func translate(text: String, to targetLanguage: String) async throws -> String {
        let systemPrompt = """
        You are a translator. Translate the following text to \(targetLanguage).
        Keep the original meaning, tone, and style. Preserve formatting like line breaks.
        Return ONLY the translated text, nothing else.
        """
        return try await chatCompletion(systemPrompt: systemPrompt, userMessage: text)
    }

    // MARK: - OpenAI Chat Completion

    private func chatCompletion(systemPrompt: String, userMessage: String) async throws -> String {
        guard !apiKey.isEmpty else { throw LLMError.noApiKey }

        let base = baseURL.hasSuffix("/") ? String(baseURL.dropLast()) : baseURL
        guard let url = URL(string: "\(base)/chat/completions") else {
            throw LLMError.invalidURL(baseURL)
        }

        let body: [String: Any] = [
            "model": model,
            "messages": [
                ["role": "system", "content": systemPrompt],
                ["role": "user", "content": userMessage]
            ],
            "temperature": 0.1,
            "max_tokens": 4096
        ]

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw LLMError.networkError(error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw LLMError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            let errorText = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw LLMError.apiError(statusCode: httpResponse.statusCode, message: errorText)
        }

        struct LLMResponse: Codable {
            struct Choice: Codable {
                struct Message: Codable {
                    let content: String?
                }
                let message: Message
            }
            let choices: [Choice]
        }

        do {
            let result = try decoder.decode(LLMResponse.self, from: data)
            guard let content = result.choices.first?.message.content, !content.isEmpty else {
                throw LLMError.emptyResponse
            }
            return content.trimmingCharacters(in: .whitespacesAndNewlines)
        } catch let error as LLMError {
            throw error
        } catch {
            throw LLMError.decodingError
        }
    }

    static func validate(baseURL: String, model: String, apiKey: String) async -> Bool {
        let base = baseURL.hasSuffix("/") ? String(baseURL.dropLast()) : baseURL
        guard let url = URL(string: "\(base)/chat/completions") else { return false }

        let body: [String: Any] = [
            "model": model,
            "messages": [["role": "user", "content": "test"]],
            "max_tokens": 1
        ]

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            return ((response as? HTTPURLResponse)?.statusCode ?? 500) < 500
        } catch {
            return false
        }
    }
}

// MARK: - Errors
enum LLMError: Error, LocalizedError {
    case noApiKey; case invalidURL(String); case invalidResponse
    case networkError(String); case apiError(statusCode: Int, message: String)
    case decodingError; case emptyResponse

    var errorDescription: String? {
        switch self {
        case .noApiKey: return "LLM API key not set. Configure in Settings."
        case .invalidURL(let u): return "Invalid Base URL: \(u)"
        case .invalidResponse: return "Invalid response from LLM API."
        case .networkError(let msg): return "Network error: \(msg)"
        case .apiError(let code, let msg): return "LLM Error (\(code)): \(msg)"
        case .decodingError: return "Failed to decode LLM response."
        case .emptyResponse: return "LLM returned empty response."
        }
    }
}
